proposals/stdlib/random.md

```diff
@@ -0,0 +1,247 @@
+* Scala: [`scala.util.Random`](https://www.scala-lang.org/api/current/scala/util/Random.html)
+* Apache commons: [RandomGenerator](http://commons.apache.org/proper/commons-math/apidocs/org/apache/commons/math4/random/RandomGenerator.html)
+* [Apache RNG](https://commons.apache.org/proper/commons-rng/userguide/rng.html): 
+[UniformRandomProvider](https://commons.apache.org/proper/commons-rng/commons-rng-client-api/apidocs/org/apache/commons/rng/UniformRandomProvider.html)
+
+## Motivation and use cases
+
+While there's no problem with generating random numbers in Kotlin/JVM (aside that `ThreadLocalRandom`
+isn't available prior to JDK 7), there's no API yet to do that in Kotlin for other targets, 
+and more importantly there's no way to do that consistently in multiplatform code.
+
+The use cases of random API are established, but here we name a few:
+
+* Sampling a random element from a collection
+* Collection shuffling
+* Random game content and level generation
+* Simulations requiring a source of random numbers
+* Using randomness in tests to cover more code execution paths
+
+## Alternatives
+
+The current alternatives require finding some platform libraries and providing common multiplatform API
+for them or implementing random generators by hand. 
+
+## Placement
+
+This API shall be placed into the Kotlin Standard Library. Regarding the package the following options were considered:
+
+- Use the existing `kotlin` package which is imported by default.
+- Use the existing `kotlin.math` package
+- Create new `kotlin.random` package and place API there.
+
+We decided to stick with the latter as it would be more explorable in docs: all the related API will be shown close to each other.
+Also it's more future proof in regard to possible API additions.
+
+A downside of a new package is that it will require new imports, especially if there will be extensions.
+A class named `Random` from that package can be confused with `java.util.Random` in the completion list.
+It should be investigated whether we could make that package imported by default in Kotlin 1.3.
+
+## Reference implementation
+
+The implementation will be available for preview in 1.3-M1.
+
+## Dependencies
+
+What are the dependencies of the proposed API:
+
+* JDK: `java.util.Random` in JDK 6,7, `ThreadLocalRandom` in JDK 8+
+* JS: `Math.random` function as the source of randomness seed
+
+## API details
+
+The following code snippet summarizes the public API introduced by this proposal.
+
+```kotlin
+abstract class Random {
+
+    abstract fun nextBits(bitCount: Int): Int
+
+    open fun nextInt(): Int 
+    open fun nextInt(bound: Int): Int 
+    open fun nextInt(origin: Int, bound: Int): Int
+    open fun nextInt(range: IntRange): Int
+
+    open fun nextLong(): Long
+    open fun nextLong(bound: Long): Long
+    open fun nextLong(origin: Long, bound: Long): Long
```

Comment:
I find the parameter names `origin` and `bound` are very non-intuitive. I kinda guess the mean `min` and `max`, but I still have no idea if they are inclusive or exclusive. I don't think we should copy obscure parameter names from JDK.

IMHO the most intuitive naming would be the same as for Kotlin ranges: `start` and `endInclusive`

Reply:
`bound` is an exclusive end, something that shouldn't be crossed. This might look unclear in the API summary, but I hope that the method [documentation](https://github.com/JetBrains/kotlin/blob/720298b53302797788778a16cd95dc43b7689e74/libraries/stdlib/src/kotlin/random/Random.kt#L53) will clarify that. If it won't, we can reconsider the parameter names.

Reply:
Ok, then it's `diapasonStart` and `diapasonEndExclusive`. You see what I mean?

Reply:
`origin` is really a bit confusing. It may give an impression that this parameter is maybe somehow used as the source of randomness instead of the internal state of the class, or that the returned value is tied to this parameter value (more than to `bound`) in some other sense.

Reply:
I've recorded the question about parameter names.

> Ok, then it's diapasonStart and diapasonEndExclusive

While we have other places in the library, where we specify half-open range with two parameters, we haven't used `exclusive` word in identifiers yet: usually it's just `startIndex`/`endIndex`